plutus-core/executables/debugger/Debugger.hs

```diff
@@ -0,0 +1,157 @@
+                , B.padTop (B.Pad 1) . BC.hCenter $ B.str cursorPos
+                ]
+    sourceEditor =
+        BB.borderWithLabel (B.txt "Source program") $
+            B.withFocusRing
+                focusRing
+                (BE.renderEditor (B.txt . Text.unlines))
+                (st ^. dsSourceEditor)
+    returnValueEditor =
+        BB.borderWithLabel (B.txt "UPLC value being returned") $
+            B.withFocusRing
+                focusRing
+                (BE.renderEditor (B.txt . Text.unlines))
+                (st ^. dsReturnValueEditor)
+    cekStateEditor =
+        BB.borderWithLabel (B.txt "CEK machine state") $
+            B.withFocusRing
+                focusRing
+                (BE.renderEditor (B.txt . Text.unlines))
+                (st ^. dsCekStateEditor)
+    ui =
+        B.vBox
+            [ BC.center uplcEditor B.<+> BC.center sourceEditor
+            , B.vLimit (st ^. dsVLimitBottomEditors) $
+                BC.center returnValueEditor B.<+> BC.center cekStateEditor
+            , footer
+            ]
+    drawTexts :: [Text] -> B.Widget ResourceName
+    drawTexts =
+        B.vBox
+            . toWidgets
+            -- This is needed for empty lines to be rendered correctly.
+            -- `Brick.Widgets.Core.txt` does the same via `fixEmpty`.
+            . fmap (\t -> if Text.null t then " " else t)
+
+    toWidgets :: [Text] -> [B.Widget ResourceName]
+    toWidgets rows = case st ^. dsUplcHighlight of
+        Nothing -> B.txt <$> rows
+        Just (HighlightCode (B.Location (sRow, sCol)) eLoc) ->
+            let alg :: (Text, Int) -> B.Widget ResourceName
+                alg (row, rowNo)
+                    | rowNo < sRow || rowNo > eRow = B.txt row
+                    -- The current line (either entirely or part of it) needs to be highlighted.
+                    | otherwise =
+                        let s =
+                                if rowNo > sRow
+                                    then 0
+                                    else sCol - 1
+                            e =
+                                if rowNo < eRow
+                                    then Text.length row
+                                    else eCol
+                            -- left -> no highlight
+                            -- middle -> highlight
+                            -- right -> no highlight
+                            ((left, middle), right) = first (Text.splitAt s) (Text.splitAt e row)
+                         in B.hBox $
+                                [B.txt left | not (Text.null left)]
+                                    ++ [ B.withAttr highlightAttr $ B.txt middle
+                                       | not (Text.null middle)
+                                       ]
+                                    ++ [B.txt right | not (Text.null right)]
+                  where
+                    B.Location (eRow, eCol) = fromMaybe (B.Location (sRow, Text.length row)) eLoc
```

Comment:
Took me a while to find this... have I mentioned I don't like `where`? :sweat_smile:

Reply:
`where` can be convenient when using pattern guards and some bindings are used by multiple conditions.